### ISSUE TITLE
AAMVA records for users with multiple middle names

### ISSUE THREAD

**Issue by RoChess** [c0shea/IdParser#30]
NameParser.cs has a logic flaw on surname for users with more than one middle names.

https://github.com/c0shea/IdParser/blob/1002a17d267b0789014703bec5b2513a28a8adb7/IdParser/Parsers/Id/NameParser.cs#L43

Suggested code fix:

`IdCard.Name.Last = names.LastOrDefault().Trim().ReplaceEmptyWithNull();`

I would also recommend the line above to become:

`IdCard.Name.Middle = string.Join(" ", names.Skip(1).Take(names.Length - 2)).Trim().ReplaceEmptyWithNull();`

This will take any middle names and merge them back together as a string separated by spaces. Retaining existing code for .Middle would work, and you could create a new IdCard field for FullMiddle or something, but the .Last one fix is needed for sure to avoid surname issues.

**Comment by c0shea** [comment]
Do you have a full example? I'd like to make sure it's for AAMVA 2000 and that the field uses a non-standard separator (like space instead of `,`, `$`, or `@`. Plus the example will help to add a unit test for this case.

**Comment by RoChess** [comment]
Dealing with HIPAA/FERPA, so unable to share the big data side. It's easy to manually recreate obviously: "John One Two Three Doe".

Having access to large data on my end processing dozens of new IDs a day has already revealed a problem with the suggested solution, as I encountered a "John Doe II" entry.

Will probably be forced to special case those.

**Comment by c0shea** [comment]
I should have more time in the next few days to take a deeper look at this. I think it will take a bit more work to handle suffixes correctly in the process.

**Comment by RoChess** [comment]
It's not easy with all the different state handling. The best success we've had so far was using an English based human nameparser project, parse the results and upon failure run it through a more agnostic variation that handles non-English names better.

Plenty of fringe cases still get caught, but at an acceptable rate for us to handle those cases via manual verification.

Especially on people with four parts to their names it gets crazy. Hopefully the Real-ID standards will be implemented by states to their full potential, because there are plenty of designed AAMVA fields to use, but many states fail to use them.

I'll try to find a way to get some data for you while remaining HIPAA/FERPA compliant. In the end you are after syntax, not content to build up use cases for testing.

**Comment by c0shea** [comment]
I found some examples in my dataset, too. I think you're right that the only thing that's going to solve the problem is when jurisdictions adopt the newer standard versions that split out each name part into its own field.

I also did some more research and found in the D20 Data Dictionary that first and last names are not supposed to contain spaces, only middle names. As a result, I've added [this commit](835b61eb56ca5f3e826e7eda2b9ca369048ec801) that does a better job of handling multiple middle names and a suffix.

It's not perfect, and jurisdictions that don't adhere to any part of the standard will continue to fail, but I don't think there's anything else I can do. Short of turning it into a natural language parser, which I don't want to, I don't see any other ways of handling those edge cases.
